File: examples/typescript/servers/mcp/simple.ts
/**
 * MCP Server with x402 Paid Tools - Simple Example
 *
 * This example demonstrates creating an MCP server with payment-wrapped tools.
 * Uses the createPaymentWrapper function to add x402 payment to individual tools.
 *
 * Run with: pnpm dev
 */

import { config } from "dotenv";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { SSEServerTransport } from "@modelcontextprotocol/sdk/server/sse.js";
import { ExactEvmScheme } from "@x402/evm/exact/server";
import { createPaymentWrapper, x402ResourceServer } from "@x402/mcp";
import { HTTPFacilitatorClient } from "@x402/core/server";
import express from "express";
import { z } from "zod";

config();

const evmAddress = process.env.EVM_ADDRESS as `0x${string}`;
if (!evmAddress) {
  console.error("❌ EVM_ADDRESS environment variable is required");
  process.exit(1);
}

const facilitatorUrl = process.env.FACILITATOR_URL;
if (!facilitatorUrl) {
  console.error("❌ FACILITATOR_URL environment variable is required");
  process.exit(1);
}

const port = parseInt(process.env.PORT || "4022", 10);

/**
 * Simulates fetching weather data for a city.
 *
 * @param city - The city name to get weather for
 * @returns Weather data object
 */
function getWeatherData(city: string): { city: string; weather: string; temperature: number } {
  const conditions = ["sunny", "cloudy", "rainy", "snowy", "windy"];
  const weather = conditions[Math.floor(Math.random() * conditions.length)];
  const temperature = Math.floor(Math.random() * 40) + 40;
  return { city, weather, temperature };
}

/**
 * Main entry point - demonstrates the payment wrapper API.
 *
 * @returns Promise that resolves when server is running
 */
export async function main(): Promise<void> {
  console.log("\n📦 Using Payment Wrapper API\n");

  // ========================================================================
  // STEP 1: Create standard MCP server
  // ========================================================================
  const mcpServer = new McpServer({
    name: "x402 Weather Service",
    version: "1.0.0",
  });

  // ========================================================================
  // STEP 2: Set up x402 resource server for payment handling
  // ========================================================================
  const facilitatorClient = new HTTPFacilitatorClient({ url: facilitatorUrl });
  const resourceServer = new x402ResourceServer(facilitatorClient);
  resourceServer.register("eip155:84532", new ExactEvmScheme());
  await resourceServer.initialize();

  // ========================================================================
  // STEP 3: Build payment requirements
  // ========================================================================
  const weatherAccepts = await resourceServer.buildPaymentRequirements({
    scheme: "exact",
    network: "eip155:84532",
    payTo: evmAddress,
    price: "$0.001",
    extra: { name: "USDC", version: "2" }, // EIP-712 domain parameters
  });

  // ========================================================================
  // STEP 4: Create payment wrapper with accepts array
  // ========================================================================
  const paidWeather = createPaymentWrapper(resourceServer, {
    accepts: weatherAccepts,
  });

  // ========================================================================
  // STEP 5: Register tools using native McpServer.tool() API
  // ========================================================================

  // Paid tool - wrap handler with payment
  mcpServer.tool(
    "get_weather",
    "Get current weather for a city. Requires payment of $0.001.",
    { city: z.string().describe("The city name to get weather for") },
    paidWeather(async (args: { city: string }) => ({
      content: [
        {
          type: "text" as const,
          text: JSON.stringify(getWeatherData(args.city), null, 2),
        },
      ],
    })),
  );

  // Free tool - no wrapper needed
  mcpServer.tool("ping", "A free health check tool", {}, async () => ({
    content: [{ type: "text", text: "pong" }],
  }));

  // Start Express server for SSE transport
  startExpressServer(mcpServer, port);
}

/**
 * Helper to start Express SSE server
 *
 * @param mcpServer - The MCP server instance
 * @param port - Port to listen on
 */
function startExpressServer(mcpServer: McpServer, port: number): void {
  const app = express();
  const transports = new Map<string, SSEServerTransport>();

  app.get("/sse", async (req, res) => {
    console.log("📡 New SSE connection");
    const transport = new SSEServerTransport("/messages", res);
    const sessionId = crypto.randomUUID();
    transports.set(sessionId, transport);
    res.on("close", () => {
      console.log("📡 SSE connection closed");
      transports.delete(sessionId);
    });
    await mcpServer.connect(transport);
  });

  app.post("/messages", express.json(), async (req, res) => {
    const transport = Array.from(transports.values())[0];
    if (!transport) {
      res.status(400).json({ error: "No active SSE connection" });
      return;
    }
    await transport.handlePostMessage(req, res, req.body);
  });

  app.get("/health", (_, res) => {
    res.json({ status: "ok", tools: ["get_weather (paid: $0.001)", "ping (free)"] });
  });

  app.listen(port, () => {
    console.log(`🚀 x402 MCP Server running on http://localhost:${port}`);
    console.log(`\n📋 Available tools:`);
    console.log(`   - get_weather (paid: $0.001)`);
    console.log(`   - ping (free)`);
    console.log(`\n🔗 Connect via SSE: http://localhost:${port}/sse`);
    console.log(`\n💡 This example uses createPaymentWrapper() to add payment to tools.\n`);
  });
}

File: examples/python/servers/mcp/simple.py
"""MCP Server with x402 Paid Tools - Simple Example.

This example demonstrates creating an MCP server with payment-wrapped tools
using the REAL MCP SDK (mcp package from PyPI).
Uses the create_payment_wrapper function to add x402 payment to individual tools.

Run with: python main.py simple
"""

import json
import os
import random
import sys
from typing import Any

from dotenv import load_dotenv
from mcp.server.fastmcp import Context, FastMCP
from mcp.types import CallToolResult

from x402.http import FacilitatorConfig, HTTPFacilitatorClientSync
from x402.mechanisms.evm.exact import ExactEvmServerScheme
from x402.mcp import (
    MCPToolResult,
    ResourceInfo,
    SyncPaymentWrapperConfig,
    create_payment_wrapper_sync,
    wrap_fastmcp_tool_sync,
)
from x402.schemas import ResourceConfig
from x402.server import x402ResourceServerSync

load_dotenv()

EVM_ADDRESS = os.getenv("EVM_ADDRESS")
FACILITATOR_URL = os.getenv("FACILITATOR_URL", "https://x402.org/facilitator")
PORT = int(os.getenv("PORT", "4022"))

if not EVM_ADDRESS:
    print("EVM_ADDRESS environment variable is required")
    sys.exit(1)

if not FACILITATOR_URL:
    print("FACILITATOR_URL environment variable is required")
    sys.exit(1)


def get_weather_data(city: str) -> dict[str, Any]:
    """Simulate fetching weather data for a city."""
    conditions = ["sunny", "cloudy", "rainy", "snowy", "windy"]
    weather = random.choice(conditions)
    temperature = random.randint(40, 80)
    return {"city": city, "weather": weather, "temperature": temperature}


def run_simple() -> None:
    """Main entry point - demonstrates the payment wrapper API with REAL MCP SDK."""
    print("\n Using Payment Wrapper API with REAL MCP SDK\n")

    # ========================================================================
    # STEP 1: Create REAL MCP server using FastMCP
    # ========================================================================
    mcp_server = FastMCP("x402-mcp-server", host="0.0.0.0", port=PORT)

    # ========================================================================
    # STEP 2: Set up x402 resource server for payment handling
    # ========================================================================
    facilitator_client = HTTPFacilitatorClientSync(FacilitatorConfig(url=FACILITATOR_URL))
    resource_server = x402ResourceServerSync(facilitator_client)
    resource_server.register("eip155:84532", ExactEvmServerScheme())
    resource_server.initialize()

    # ========================================================================
    # STEP 3: Build payment requirements
    # ========================================================================
    accepts = resource_server.build_payment_requirements(
        ResourceConfig(
            scheme="exact",
            network="eip155:84532",
            pay_to=EVM_ADDRESS,
            price="$0.001",
            extra={"name": "USDC", "version": "2"},
        )
    )

    # ========================================================================
    # STEP 4: Create payment wrapper with accepts array
    # ========================================================================
    paid_weather = create_payment_wrapper_sync(
        resource_server,
        SyncPaymentWrapperConfig(accepts=accepts),
    )

    # ========================================================================
    # STEP 5: Register tools - wrap handler with payment
    # ========================================================================

    # Create the paid tool bridge (SDK handles _meta extraction + CallToolResult conversion)
    paid_weather_tool = wrap_fastmcp_tool_sync(
        paid_weather,
        lambda args, _: MCPToolResult(
            content=[{"type": "text", "text": json.dumps(get_weather_data(args["city"]))}],
        ),
        tool_name="get_weather",
    )

    # Paid tool - just call the bridge with args + context
    @mcp_server.tool()
    def get_weather(city: str, ctx: Context) -> CallToolResult:
        """Get current weather for a city. Requires payment of $0.001."""
        return paid_weather_tool({"city": city}, ctx)

    # Free tool - no wrapper needed
    @mcp_server.tool()
    def ping() -> str:
        """A free health check tool."""
        return "pong"

    # Run server with SSE transport (interoperable with Go/TypeScript clients)
    print(f"x402 MCP Server running on http://localhost:{PORT}")
    print("\nAvailable tools:")
    print("   - get_weather (paid: $0.001)")
    print("   - ping (free)")
    print(f"\nConnect via SSE: http://localhost:{PORT}/sse\n")

    mcp_server.run(transport="sse")
